src/core/hle/kernel/thread.cpp

```diff
@@ -3,433 +3,972 @@
+        context->SetCpuRegister(1, out_value);
     }
+    return result;
+}
 
-    if (thread->status == ThreadStatus::WaitSynchAny ||
-        thread->status == ThreadStatus::WaitSynchAll || thread->status == ThreadStatus::WaitArb ||
-        thread->status == ThreadStatus::WaitHleEvent) {
+ResultCode ThreadManager::WaitOne(ObjectPtr object, std::chrono::nanoseconds timeout) {
+    ASSERT(thread->status == Thread::Status::Running);
 
-        // Invoke the wakeup callback before clearing the wait objects
-        if (thread->wakeup_callback)
-            thread->wakeup_callback->WakeUp(ThreadWakeupReason::Timeout, thread, nullptr);
+    // Check if we can sync immediately
+    if (!object->ShouldWait(thread)) {
+        object->Acquire(thread);
+        return thread->SyncOutput(RESULT_SUCCESS);
+    }
 
-        // Remove the thread from each of its waiting objects' waitlists
-        for (auto& object : thread->wait_objects)
-            object->RemoveWaitingThread(thread.get());
-        thread->wait_objects.clear();
+    // Schedule the timeout
+    if (timeout.count() == 0) {
+        return thread->SyncOutput(RESULT_TIMEOUT);
+    } else {
+        root.ScheduleEvent(thread->wakeup_event.get(), timeout);
     }
 
-    thread->ResumeFromWait();
+    // Wait on the object
+    thread->waiting_on = {object};
+    object->AddWaitingThread(Kernel::SharedFrom(thread));
+    thread->status = Thread::Status::WaitSyncAll;
+    Reschedule();
+    return thread->SyncOutput(RESULT_SUCCESS);
 }
 
-void Thread::WakeAfterDelay(s64 nanoseconds) {
-    // Don't schedule a wakeup if the thread wants to wait forever
-    if (nanoseconds == -1)
-        return;
+ResultCode ThreadManager::WaitAny(std::vector<ObjectPtr> objects, std::chrono::nanoseconds timeout,
+                                  s32* index) {
+    ASSERT(thread->status == Thread::Status::Running);
+    thread->sync_output = index;
+
+    // Check if we can sync immediately
+    auto active_obj = std::find_if(objects.begin(), objects.end(),
+                                   [this](const auto& obj) { return !obj->ShouldWait(thread); });
+    if (active_obj != objects.end()) {
+        (*active_obj)->Acquire(thread);
+        return thread->SyncOutput(RESULT_SUCCESS, static_cast<u32>(active_obj - objects.begin()));
+    }
 
-    thread_manager.kernel.timing.ScheduleEvent(nsToCycles(nanoseconds),
-                                               thread_manager.ThreadWakeupEventType, thread_id);
+    // Schedule the timeout
+    if (timeout.count() == 0) {
+        auto inactive_obj = std::find_if(objects.begin(), objects.end(), [this](const auto& obj) {
+            return obj->ShouldWait(thread);
+        });
+        if (inactive_obj != objects.end()) {
+            return thread->SyncOutput(RESULT_TIMEOUT);
+        }
+    } else {
+        root.ScheduleEvent(thread->wakeup_event.get(), timeout);
+    }
+
+    thread->waiting_on = objects;
+    for (auto obj : objects) {
```

Comment:
```suggestion
    for (auto& obj : objects) {
```

otherwise this can cause atomic reference count increment/decrement churn.